ksql-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java

```diff
@@ -20,72 +20,115 @@
 import io.confluent.ksql.parser.tree.Expression;
 import io.confluent.ksql.parser.tree.FunctionCall;
 import io.confluent.ksql.parser.tree.Node;
-import io.confluent.ksql.parser.tree.QualifiedName;
-import io.confluent.ksql.parser.tree.QualifiedNameReference;
-import io.confluent.ksql.util.SchemaUtil;
+import io.confluent.ksql.util.KsqlException;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Stack;
+import java.util.function.BiConsumer;
 
-public class AggregateAnalyzer extends DefaultTraversalVisitor<Node, AnalysisContext> {
+class AggregateAnalyzer {
 
-  private final AggregateAnalysis aggregateAnalysis;
-  private final Analysis analysis;
+  private final MutableAggregateAnalysis aggregateAnalysis;
+  private final DereferenceExpression defaultArgument;
   private final FunctionRegistry functionRegistry;
 
-  private boolean hasAggregateFunction = false;
+  AggregateAnalyzer(
+      final MutableAggregateAnalysis aggregateAnalysis,
+      final DereferenceExpression defaultArgument,
+      final FunctionRegistry functionRegistry
+  ) {
+    this.aggregateAnalysis = Objects.requireNonNull(aggregateAnalysis, "aggregateAnalysis");
+    this.defaultArgument = Objects.requireNonNull(defaultArgument, "defaultArgument");
+    this.functionRegistry = Objects.requireNonNull(functionRegistry, "functionRegistry");
+  }
+
+  void processSelect(final Expression expression) {
+    final Set<DereferenceExpression> nonAggFields = new HashSet<>();
+    final AggregateVisitor visitor = new AggregateVisitor((aggFuncName, node) -> {
+      if (!aggFuncName.isPresent()) {
+        nonAggFields.add(node);
+      }
+    });
 
-  public boolean isHasAggregateFunction() {
-    return hasAggregateFunction;
+    visitor.process(expression, new AnalysisContext());
+
+    if (!visitor.visitedAggFunction) {
```

Comment:
You could have a non-group-by column in the expression if its outside the aggregation function:
```
SELECT X, Y + COUNT(*) GROUP BY X
```
Maybe makes sense to call `addNonAggregateSelectExpression` from the collector

Reply:
Is something like this allowed?

Reply:
I don't see why not if we're grouping by the non-aggregate column. Its potentially useful too. For example, if someone is tracking some resource with an up-front and per-unit cost that are represented as other columns in their schema they could do something like:

```SELECT RESOURCE, UP_FRONT + PER_UNIT * COUNT(*) AS TOTAL_COST FROM USAGE GROUP BY RESOURCE, UP_FRONT, PER_UNIT```

That kind of query does build/execute today.

I think the case described in my original example should not be allowed since the column from the expression isn't in the group by. That said, this is kind of orthogonal to what you're doing in this PR (enhancing what you can put in group-by). On the other hand it's a small change that I think tightens up the semantics. Your call.

Reply:
> I think the case described in my original example should not be allowed since the column from the expression isn't in the group by

Yeah, that's what confused me.

I'll take a look at supporting the second example you've supplied.

Reply:
I've added tests to cover this and the code looks to be 'doing the right thing' to me.

The expression `Y + COUNT(*)` would still set `visitedAggFunction` to true. Which is what we want.

Reply:
I was saying we would want it to error out. Since it’s using a column that’s omitted from GROUP BY

Reply:
I was saying we would want it to error out. Since it’s using a column that’s omitted from GROUP BY

Reply:
I was saying we would want it to error out. Since it’s using a column that’s omitted from GROUP BY

Reply:
Sorry, I mean `UP_FRONT + PER_UNIT * COUNT(*)` would set it to true and do the right thing.

I'll add a test to cover `Y + COUNT(*)`